topic_dynamics/parsing.py

```diff
@@ -0,0 +1,342 @@
+    tree = get_parser(lang).parse(code)
+    root = tree.root_node
+    identifiers = []
+
+    def traverse_tree(node: tree_sitter.Node) -> None:
+        """
+        Run down the AST (DFS) from a given node and gather identifiers from its children.
+        :param node: starting node.
+        :return: None.
+        """
+        for child in node.children:
+            if child.type in NODE_TYPES[lang]:
+                start, end = get_positional_bytes(child)
+                identifier = code[start:end].decode("utf-8").lower()
+                if "\n" not in identifier:  # Will break output files. Can add other bad characters later
+                    identifiers.append(identifier)
+            if len(child.children) != 0:
+                traverse_tree(child)
+
+    traverse_tree(root)
+    sorted_identifiers = sorted(Counter(identifiers).items(), key=itemgetter(1), reverse=True)
+
+    return sorted_identifiers
+
+
+def transform_identifiers(identifiers: List) -> List[str]:
+    """
+    Transform the original list of identifiers into the writable form.
+    :param identifiers: list of tuples, identifier and count.
+    :return: a list of identifiers in the writable form, "identifier:count".
+    """
+    formatted_identifiers = []
+    for identifier in identifiers:
+        if identifier[0].rstrip() != "":  # Checking for occurring empty tokens.
+            formatted_identifiers.append(identifier[0].rstrip() + ":" + str(identifier[1]).rstrip())
+    return formatted_identifiers
+
+
+def slice_and_parse(repository: str, n_dates: int, time_delta: int, lang: str, name: str) -> None:
+    """
+    Split the repository, parse the full files, write the data into a file.
+    Can be called for parsing full files and for parsing diffs only.
+    :param repository: path to the repository to process.
+    :param n_dates: the amount of dates.
+    :param time_delta: the time step between dates.
+    :param lang: language of parsing.
+    :param name: name of the dataset (directories with resulting files).
+    :return: None.
+    """
+    print("Creating the temporal slices of the data.")
+    # Create a folder for created files
+    directory = os.path.abspath(os.path.join(repository, os.pardir, name + "_processed"))
+    os.mkdir(directory)
+    dates = get_dates(n_dates, time_delta)
+    dates_indices = {}
+    count = 0
+    # Create temporal slices of the project, get a list of files for each slice, parse all files, save the tokens
+    with open(os.path.abspath(os.path.join(directory, name + "_tokens.txt")), "w+") as fout:
+        for date in tqdm(dates):
+            with TemporaryDirectory() as td:
+                subdirectory = os.path.abspath(os.path.join(directory, td, date.strftime("%Y-%m-%d")))
+                checkout_by_date(repository, subdirectory, date)
+                files = get_files(subdirectory, get_extensions(lang))
+                starting_index = count + 1
+                for file in files:
```

Comment:
what will happen if none of the files in slice will have tokens? It will create slice with no elements, and next slice will start with next element - logic should be fixed

Reply:
Added this check to both full files and diffs. It checks that ending index >= starting index. If that is not the case (meaning the slice was empty), it is skipped.